storage/blockdevice/source/SFDP.cpp

```diff
@@ -291,13 +291,96 @@ int sfdp_parse_sector_map_table(Callback<int(bd_addr_t, sfdp_cmd_addr_size_t, ui
         return -1;
     }
 
-    // Currently we support only Single Map Descriptor
-    if (!((smptbl_buff[0] & 0x3) == 0x03) && (smptbl_buff[1] == 0x0)) {
-        tr_error("Sector Map: Supporting Only Single Map Descriptor (not map commands)");
+    // Loop through all descriptors: configuration command descriptor and sector map descriptors.
+    // Command descriptors are optional. If any exist, they always precede sector map descriptors.
+    uint8_t *table = smptbl_buff.get();
+    uint8_t *descriptor = table;
+    const size_t min_descriptor_size = 8; // two DWORDs
+
+    // Part 1: configuration detection command descriptors (optional).
+    // If any exist, run their commands to determine the active configuration's ID.
+    // Otherwise the configuration ID is 0x00.
+    uint8_t active_config_id = 0x00;
+    // If any configuration detection command descriptors exist, the table starts with one.
+    if (!(descriptor[0] & 0x02)) {
+        while (descriptor + min_descriptor_size <= table + sfdp_info.smptbl.size) { // table boundary
+            if (descriptor[0] & 0x02) {
+                tr_error("Sector Map: Expecting a configuration detection command descriptor");
+                return -1;
+            }
+
+            uint8_t instruction = descriptor[1];
+            uint8_t dummy_cycles = descriptor[2] & 0x0F;
+            auto addr_size = static_cast<sfdp_cmd_addr_size_t>(descriptor[2] >> 6);
+            uint8_t mask = descriptor[3];
+            uint32_t cmd_addr;
+            memcpy(&cmd_addr, &descriptor[4], sizeof(cmd_addr)); // bytes 4-7 of the descriptor
+
+            uint8_t rx;
+            status = sfdp_reader(cmd_addr, addr_size, instruction, dummy_cycles, &rx, sizeof(rx));
+            if (status < 0) {
+                tr_error("Sector Map: Configuration detection command failed");
+                return -1;
+            }
+
+            // Shift existing bits to the left, so we can add the newly detected bit
+            active_config_id <<= 1;
+
+            // The mask may apply to any bit of rx, so we can't directly combine
+            // (rx & mask) with active_config_id. Instead, treat (rx & mask) as a boolean.
+            if (rx & mask) {
+                active_config_id |= 0x01;
+            }
+
+            // Each command descriptor has a fixed size of two DWORDS.
+            if (descriptor[0] & 0x01) { // the current command descriptor is the last one
+                descriptor += min_descriptor_size; // next descriptor
+                break;
+            }
+            descriptor += min_descriptor_size; // next descriptor
+        }
+    }
+
+    // Part 2: sector map descriptors.
```

Comment:
Does it make sense to factor out "part one" and "part 2" into separate helper functions with clear names?

Reply:
Good point. I wanted reviewers' opinion on breaking down this big function, now you confirmed it so I'll do it.

Reply:
Done

Reply:
In commit "SFDP: Add support for multiple configurations and sector maps", I still see "Part 1" and "Part 2" in the same function.

Reply:
The code was already refactored, but I kept the comments. Now I've rewritten the comments to not say "Part 1" and "Part 2".